[Step1] 리뷰 부탁드립니다

mia6111 opened the pull request.

Commit:
feat step1 문자열 계산기 작성 및 테스트 코드 작성

Commit:
refactor step1 요구사항 잘못이해한 부분 및 StringSplitStrategy 구조 변경

Commit:
style step1 줄바꿈/변수명 변경/오타 픽스

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/CustomStringSplitStrategy.java:
```diff
@@ -0,0 +1,29 @@
+    private static final Pattern CUSTOM_IGNORE_REGEX = Pattern.compile("(//.*\n)");
+    private final Pattern splitRegex;
+
+    CustomStringSplitStrategy(String... splitStrings) {
```
저는 자료형을 메서드명 혹은 변수명에 사용하는 것은 지양하고 있습니다.

https://github.com/Yooii-Studios/Clean-Code/blob/master/Chapter%2002%20-%20의미%20있는%20이름.md#의미-있게-구분하라불용어-noise-word-를-쓰지-말자

brainbackdoor commented on src/test/java/camp/nextstep/edu/calculator/CustomStringSplitStrategyTest.java:
```diff
@@ -0,0 +1,53 @@
+        return Stream.of(
+            arguments(new String[]{";"}, "a;b", Arrays.asList("a","b")),
+            arguments(new String[]{";", "!"}, "a;b!c", Arrays.asList("a","b","c")),
+            arguments(new String[]{"|", "~"}, "ab|cd~ef",  Arrays.asList("ab","cd","ef")),
```
이 부분에 대한 요구사항은 없었던 거 같아요

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,56 @@
+    public static PositiveNumber parse(String input) {
+        try {
+            return new PositiveNumber(Integer.parseInt(input));
+        }catch (NumberFormatException e){
```
공백 컨벤션 위반입니다. IDE에서 컨벤션을 자동으로 맞춰주는 shortcut이 있으니 확인해보세요.
intelliJ 는 option + command + l 입니다.

8.2 blank spaces 항목을 확인해보세요.
https://myeonguni.tistory.com/1596

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,56 @@
+        try {
+            return new PositiveNumber(Integer.parseInt(input));
+        }catch (NumberFormatException e){
+            throw new IllegalArgumentException("input must be number");
```
exception을 흘려보내는 것은 Anti pattern으로 알고 있어요. 
아래 링크에 Exception 처리에 대해서 잘 정리되어있으니 이번 기회에 같이 확인해보는 것도 좋겠네요 ㅎ

https://www.slipp.net/questions/350

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,56 @@
+
+    public static PositiveNumber parse(Integer input) {
+        return new PositiveNumber(input.longValue());
+    }
```
위의 정적 팩토리 메서드는 테스트를 위한 것으로 보이는데요.
저는 테스트만을 위한 production 코드를 작성하는 것은 지양하고 있습니다.
테스트의 성공이 실제 프로덕션 코드에서 의도했던 바를 성공시킨 것인지 확신할 수 없기 때문입니다

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,56 @@
+        if (CollectionUtils.isEmpty(inputList)) {
+            return Collections.emptyList();
+        }
+        return inputList.stream().map(PositiveNumber::parse).collect(toList());
```
List, Set 등 자료구조를 지칭하는 단어를 클래스나 변수이름으로 사용하는 것을 지양하고 있습니다.
사용하는 자료구조가 변경되면 변경의 범위가 외부로 전파될 수 있기 때문입니다.
List와 같은 복수의 데이터의 경우 복수형을 쓰는 것이 더 적합하다고 생각합니다.

https://github.com/Yooii-Studios/Clean-Code/blob/master/Chapter%2002%20-%20의미%20있는%20이름.md#그릇된-정보를-피하라

brainbackdoor replied:
이 메서드의 역할이 불분명하다고 생각합니다. (PositiveNumber가 List<PositiveNumber>를 반환하는 메서드를 가지는 것이 어색하지 않을지..)

그리고 메서드명은 parseString이 적절할까요 (정적팩토리 메서드의 경우 흔히 사용되는 메서드이름들이 있습니다. (valueOf, of, newInstance 등..)
https://johngrib.github.io/wiki/static-factory-method-pattern/

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,56 @@
+
+    public boolean isSameAs(Integer other){
+        return other!=null && this.value == other;
+    }
```
사용하지 않는 코드는 제거해주세요.

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,56 @@
+        this.value = value;
+    }
+
+    public PositiveNumber add(PositiveNumber next) {
+        Assert.notNull(next, "next must be not null");
+        return new PositiveNumber(this.value + next.value);
+    }
```
👍

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    private PositiveNumber sum(List<PositiveNumber> numbers) {
+        return numbers.stream().reduce((prev,next) -> prev.add(next)).get();
```
lambda식을 사용하면 method reference보다 참조가 한번 더 일어납니다.

https://softwareengineering.stackexchange.com/questions/277473/is-there-a-performance-benefit-to-using-the-method-reference-syntax-instead-of-l

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/StringSplitStrategy.java:
```diff
@@ -0,0 +1,25 @@
+    default String[] escapeCharacters(String[] inputList) {
+        return Arrays.stream(inputList).map(Pattern::quote).toArray(String[]::new);
+    }
+    class StringSplitStrategyFactory {
```
이 클래스를 sub class로 가져감으로써 어떤 이점이 있는걸까요? 🤔

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/StringSplitStrategy.java:
```diff
@@ -0,0 +1,25 @@
+
+    default String[] escapeCharacters(String[] inputList) {
+        return Arrays.stream(inputList).map(Pattern::quote).toArray(String[]::new);
+    }
```
interface의 default 메서드는 정말 특별한 경우가 아니면 사용하는 것을 지양하고 있는데요.
(원래 interface의 컨셉에도 맞지 않구요.)
무엇보다 위의 메서드는 CustomStringSplitStrategy에서만 사용하는 것이니 CustomStringSplitStrategy에 역할을 위임하는 것이 좋지 않을까요

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,20 @@
+        if(input == null || input.isEmpty()){
+            return 0L;
+        }
+        List<String> splitStrings =  StringSplitStrategyFactory.getStrategy(input).apply(input);
```
디미터 원칙 위반입니다.
classA.methodA의 최대 지식한계는 아래와 같습니다. 
- classA의 필드 객체
- methodA가 생성한 객체
- methodA의 인자로 넘어온 객체 

다른 객체 내부의 데이터가 무엇이 있는지까지 알아서는 안되며,
역할을 담당하고 있는 객체에 메시지를 보내서 응답받도록 작성해보세요.

https://developerfarm.wordpress.com/2012/01/30/object_calisthenics_5/

---

가령, 이 경우엔 아래와 같이 작성해보는 것은 어떨까요

```suggestion
        List<String> splited =  StringUtils.split(input);
```

brainbackdoor commented on src/test/java/camp/nextstep/edu/calculator/PatternTest.java:
```diff
@@ -0,0 +1,27 @@
+    }
+    @Test
+    public void compile_escaped() {
+//        Pattern compile1 = Pattern.compile("|", Pattern.LITERAL);
```
주석을 제거해주세요.

brainbackdoor commented on src/test/java/camp/nextstep/edu/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,79 @@
+        StringCalculator calculator = new StringCalculator();
+        long result = calculator.calculateSum(testString);
+        //then
+        assertThat(result).isEqualTo(0);
```
이 테스트가 실패하네요.  😰

brainbackdoor commented on src/test/java/camp/nextstep/edu/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,79 @@
+    }
+    public static Stream<Arguments> customStrategy(){
+        return Stream.of(
+            arguments("//;\\n1;2;3", 6),
```
```suggestion
            arguments("//;\n1;2;3", 6),
```

요구사항과 다른 테스트 입력값으로 보입니다.

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/CustomStringSplitStrategy.java:
```diff
@@ -0,0 +1,29 @@
+import org.thymeleaf.util.StringUtils;
+
+public class CustomStringSplitStrategy implements StringSplitStrategy {
+    public static final Pattern CUSTOM_SPLIT_REGEX = Pattern.compile("//(.*)\n");
```
```suggestion
    public static final Pattern CUSTOM_SPLIT_REGEX = Pattern.compile("//(.*)\\n(.*)");
```

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/StringSplitStrategy.java:
```diff
@@ -0,0 +1,25 @@
+        public static StringSplitStrategy getStrategy(String input) {
+            Matcher matcher = CustomStringSplitStrategy.CUSTOM_SPLIT_REGEX.matcher(input);
+            if (matcher.matches()) {
+                return new CustomStringSplitStrategy(matcher.group(0));
```
```suggestion
                return new CustomStringSplitStrategy(matcher.group(1));
```

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/DefaultStringSplitStrategy.java:
```diff
@@ -0,0 +1,21 @@
+    private static final DefaultStringSplitStrategy DEFAULT = new DefaultStringSplitStrategy();
+    private static final Pattern DEFAULT_SPLIT_PATTERN = Pattern.compile(":|,");
+
+    private DefaultStringSplitStrategy(){}
```
👍

brainbackdoor commented on src/test/java/camp/nextstep/edu/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,79 @@
+        assertThatIllegalArgumentException().isThrownBy(()->{
+            StringCalculator calculator = new StringCalculator();
+            long result = calculator.calculateSum(testString);
+        });
```
라인을 줄일 때는 기준이 명확해야 하다고 생각해요.

그리고 지역변수를 별도로 둘 때는 의미가 분명해야 한다고 생각합니다.

```java
    @DisplayName("숫자 이외 문자열 또는 음수 문자열이 주어지면 실패한다")
    @ParameterizedTest
    @ValueSource(strings = {
            "a,2",
            "!,2",
            "-100,2",
            "//;\\na;2",
            "//~\\n!:2",
            "//~\\n-100,2"
    })
    void calculate_given_invalid_number(String unexpectedInput) {
        StringCalculator calculator = new StringCalculator();

        assertThatIllegalArgumentException().isThrownBy(() -> calculator.sum(unexpectedInput));
    }
```

brainbackdoor commented on src/test/java/camp/nextstep/edu/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,79 @@
+class StringCalculatorTest {
+    @DisplayName(":를 구분자로 문자열을 파싱하여 문자열 숫자의 합을 반환한다")
+    @ParameterizedTest
+    @MethodSource("camp.nextstep.edu.calculator.StringCalculatorTestProvider#defaultStrategy")
```
```suggestion
    @MethodSource("defaultStrategy")
```

테스트 코드 중복을 제거하기 위해 분리하신거라면 상속구조로 가져가는 것은 어떨까요

brainbackdoor reviewed (changes requested):
안녕하세요! 리뷰어 이동규입니다.
테스트케이스를 상세히 나누어주셨네요! 👍 ParameteralizedTest를 적극적으로 활용하신 부분도 인상적입니다.
이번  과정을 진행하면서 commit을 기능 단위로 나누고 Commit message를 충실히 작성하는 연습을 해보는 것도 나름대로 의미가 있다고 생각합니다.
https://meetup.toast.com/posts/106
https://www.conventionalcommits.org/en/v1.0.0-beta.4/
몇가지 피드백을 남겨두었으니 확인해보시고 궁금한 점 있으면 메시지 남겨주세요.
남은 주말 잘 보내세요~ 😀

Commit:
refactor step1 요구사항 관련 없는 코드 삭제 및 static 생성자 메소드명 변경

- PositiveNumber 코드 중 요구사항 관련 없는 코드 삭제 및 static 생성자 메소드명 변경

Commit:
refactor step1 불필요하게 복잡한 Strategy 구조 삭제 및 SentenceSplitter 클래스로 로직 이관

- 꺠지는 테스트 픽스
- 불필요하게 복잡한 상속 구조 삭제 (StringSplitStrategy <- Default / CustomStringStrategy) 및 로직을 SentenceSplitter 클래스로 치환

Commit:
refactor step1 클래스/메소드/변수명 변경

Mia6111 commented:
상세한 답변 너무나도 감사합니다 ㅠㅠ 많은 시간 써서 답변 주신것 같아 송구스럽고 감사하네요.
문자열 분리하는 로직 부분을 추상화해서 오히려 복잡도가 늘어난 부분이 마음에 걸렸는데  (-디미터원칙을 위반하게 되는 부분과 팩토리까지 추가하면서 복잡..) 역시 잘못된 선택같아서 수정했습니다! 
정확하게 지적해주시고 알려주셔서 감사합니다!! 많은 도움 되었습니다.

+ 깨지는 테스트 제대로 확인 못한점과 format convention 위반과 같은 기본적인 실수한 점 죄송합니다 ㅠㅠ 다음부턴 푸쉬 전에 다시 한번 확인해보도록 하겠습니다

Commit:
style 테스트 코드 포맷팅 적용

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,48 @@
+        }
+    }
+
+    private PositiveNumber(long value) {
+        if (value < 0) {
+            throw new IllegalArgumentException("value must be equal or greater than 0");
+        }
+        this.value = value;
+    }
```
method 순서에 대해서도 한번 고민해보셔요.

https://stackoverflow.com/questions/4668218/are-there-any-java-method-ordering-conventions

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public boolean isSame(PositiveNumber other) {
+        return other != null && this.value == other.value;
```
```suggestion
        return Objects.equals(this.value, other.value);
```

Java 7부터 제공되는 Objects 클래스가 있습니다. 아래의 글을 참고하세요.
https://multifrontgarden.tistory.com/205

brainbackdoor commented on src/main/java/camp/nextstep/edu/calculator/SentenceSplitter.java:
```diff
@@ -0,0 +1,31 @@
+    private static final Pattern DEFAULT_SPLIT_PATTERN = Pattern.compile(":|,");
+    private static final Pattern CUSTOM_SPLIT_REGEX = Pattern.compile("//(.)\\n(.*)");
+
+    public static List<String> split(String input) {
+        if (StringUtils.isEmpty(input)) {
+            return Collections.emptyList();
+        }
+        Matcher matcher = CUSTOM_SPLIT_REGEX.matcher(input);
+        if (!matcher.find()) {
+            return splitByDelimiter(DEFAULT_SPLIT_PATTERN, input);
+        }
+
+        Pattern delimiter = Pattern.compile(matcher.group(1), Pattern.LITERAL);
+        String target = matcher.group(2);
+        return splitByDelimiter(delimiter, target);
```
메서드를 작은 단위로 쪼개어 책임을 분리하고 의미를 분명히 전달하는 것은 어떨까요

```java
    public static List<String> execute(String input) {
        if (isEmpty(input)) {
            return emptyList();
        }

        return split(input);
    }

    private static List<String> split(String input) {
        Matcher matcher = CUSTOM_SPLIT_REGEX.matcher(input);
        if (!matcher.find()) {
            return splitByDelimiter(DEFAULT_SPLIT_PATTERN, input);
        }

        return splitByDelimiter(extractCustomDelimiter(matcher), extractTarget(matcher));
    }
```

brainbackdoor reviewed (approved):
피드백 반영 정말 잘해주셨네요! 👍 
몇가지 간단한 피드백 남겼으니 확인해보시고, 궁금한 점 있으면 메시지 남겨주세요.
오늘 하루 마무리 잘하시고 내일 수업때 뵈어요~ 🤗
